**Group 08**

Ana Cláudia Alferes - M20190932


Francisco Carujo Neves - M20190089


Luís Filipe Pinho - M20190935

# Big Data Modeling and Management Assigment


## 🍺 The Beer project  🍺 

As it was shown in classes, graph databases are a natural way of navegating distinct types of data. For this first project we will be taking a graph database to analyse beer and breweries!   

_For reference the dataset used for this project has been extracted from [kaggle](https://www.kaggle.com/ehallmar/beers-breweries-and-beer-reviews), released by Evan Hallmark. Even though the author does not present metada on the origin of the data it is probably a collection of open data from places like [beeradvocate](https://www.beeradvocate.com/)_ 

#### Problem description

Explore the database via python neo4j connector and/or the graphical tool in the NEO4J webpage. Answer the questions. Submit the results by following the instructions

#### Connection details to the neo4j database
```
Host: rhea.isegi.unl.pt:7474  
Username: neo4j  
Password: F3cfcrnvBev57KZ8mcMk78L9wHgJVZuJ 
Connect URL : bolt://rhea.isegi.unl.pt:7687
```


#### Questions


0. __Example__ Question _How many beers does the database contain?_
1. How many countries are accounted in this database?
1. Which entity holds the most reviews:  
    1. Beer?  
    1. Brewery?
    1. Country? 
1. Find the user/users that have the most shared reviews(reviews or the same beers) with the user CTJman?
1. Which Portuguese brand has the most beers?
1. Out of those beers, which has the most reviews?
1. On average how many different beer styles does each brewery produce?
1. Which brewery produces the strongest beers according to ABV?
1. If I typically enjoy a beer due to its aroma and appearance, which beer style should I try?
1. If you had to pick 3 beers to recommend using only this data, which would you pick?

##### Groups  
Groups should have 3 to 4 people  
You should register your group in the link:
- https://liveeduisegiunl-my.sharepoint.com/:x:/g/personal/fpinheiro_novaims_unl_pt/EYtFTvuGwiRCsakM67f3C4oBrafEJ8M1juXlP7OUdNg2PA?rtime=BlYdnK_n10g


##### Submission      

Submission of the query results to be done to the redis database (explained on the first class), with the exception of question 9.  
The following format is expected:
```
    >>> redis.set("GROUP1:0", "358873")
```

This result should be the anwser of group 1 to question 0

The solutions with the code used to produce the results and respective explations should be uploaded to moodle. They should have a clear reference to the group, either on the file name or on the document itself. Preferably one Jupyter notebook per group.

Delivery date: Until the **midnight of May 7th**

##### Evaluation   

This will be 15% of the final grade.   
Each solution will be evaluate on 2 components: correctness of results and simplicity of the solution.  
All code will go through plagiarism automated checks. Groups with the same code will undergo investigation.

In [1]:
import py2neo
from pprint import pprint
import redis

Host: rhea.isegi.unl.pt:7474  
Username: neo4j  
Password: F3cfcrnvBev57KZ8mcMk78L9wHgJVZuJ

Connect URL : bolt://rhea.isegi.unl.pt:7687

In [2]:
R_edis = redis.Redis(
    host='rhea.isegi.unl.pt',                            
    port='6379',                                         
    password='4k9krMVdstxm9VvsSQQcvB7zZxwmWLqMXwxqctM5', 
    decode_responses=True
)

In [3]:
username="neo4j"
password="F3cfcrnvBev57KZ8mcMk78L9wHgJVZuJ"
host="rhea.isegi.unl.pt"
port="7474"

In [4]:
secure_graph = py2neo.Graph(f"http://{username}:{password}@{host}:{port}")
secure_graph.run("MATCH () RETURN count(*)").data()

[{'count(*)': 9647597}]

# Some Information

Explanations and submission of the answers are done directly below each query

In [5]:
##QUESTION 0

In [6]:
result = secure_graph.run("""
        MATCH (b:Beers)
        RETURN count (*)
""").data()
pprint(result)

[{'count (*)': 358873}]


In [7]:
R_edis.set("GROUP8:0", 358873)

True

**Explanation**

To answer this example question, we matched the Beers table, which contains all the beers in the database from all the breweries. Because we want to know how many beers are in this database, we needed the count of the total number of Beers, so we used RETURN count(*), and the result is 358873 beers in the database.

**Result**
358873 beers


In [8]:
##QUESTION 1

In [9]:
result = secure_graph.run("""
        MATCH (c:Country)
        RETURN count (*)
""").data()
pprint(result)

[{'count (*)': 200}]


In [10]:
R_edis.set("GROUP8:1", 200)

True

For question number 1, we want to know the total number of countries in the database. For that purpose, we matched the table Country, which contains all the countries in the database. To return the total number, we used the count(*) to count all the rows in this table. So, the total number of countries in the database is 200.

In [11]:
##QUESTION 2

In [12]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)
        RETURN b.name,count(b.name) as cnt
        ORDER BY cnt DESC
        LIMIT 1
""").data()
pprint(result)

[{'b.name': 'IPA', 'cnt': 31387}]


In [13]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)-[ba:BREWED_AT]->(br:Breweries)
        RETURN br.name,count(br.name) as cnt
        ORDER BY cnt DESC
        LIMIT 1
""").data()
pprint(result)

[{'br.name': 'Sierra Nevada Brewing Co.', 'cnt': 175161}]


In [14]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)-[ba:BREWED_AT]->(br:Breweries)-[f:FROM]->(c:Country)
        RETURN c.country_digit,count(c.country_digit) as cnt
        ORDER BY cnt DESC
        LIMIT 1
""").data()
pprint(result)

[{'c.country_digit': 'US', 'cnt': 7524410}]


In [15]:
R_edis.set("GROUP8:2A", 'IPA')

True

In [16]:
R_edis.set("GROUP8:2B", 'Sierra Nevada Brewing Co.')

True

In [17]:
R_edis.set("GROUP8:2C", 'US')

True

For question 2, we will divide our answer in three parts, because we performed three different queries.
For part A, we needed the beer with the highest number of reviews. For that reason, we matched the table Reviews, which contains all the reviews in the database. However, we only wanted the reviews on Beers, so we used the connection "About" between the tables Reviews and Beers to explain that we only needed beers. In our output, we wanted the algorithm to return the name of the beer and the number or reviews (calculated by the count of the times the name of the beer appeared). Finally, we ordered the output in descending order, in order to see in first place, the beer with the highest number of reviews, which is "IPA", with 31387 reviews.
In part B, we were requested the Brewery with the highest amount of beer reviews. To answer this question, we started the same way as in part A, by checking the beers in the table Reviews. However, in this question, we had to go deeper, in order to know to which brewery those beers belong to. So, we used the connection "BREWED_AT", to know, from the beers which were reviewed, what is the Brewery that produces them. As an output, we wanted the name of the brewery and the total number of times that the name of that brewery appeared, calculated with a count, and ordered in descending order. By looking at the output, the brewery with the highest number of reviews is Sierra Nevada Brewing Co., with 175161 reviews.
Finally, in part C, we wanted to know which is the country with the highest number of beer reviews. We calculated the same way as in the previous parts, by matching the table reviews, and to know the beers that were reviewed, and which brewery produced them. Now, with the connection "FROM" between the tables Breweries and Country, we will know from which country is every Brewery. For our output, we wanted the algorithm to return the country digit, to identify the country, and the number of times that this digit appeared in the reviews, calculated by counting the names, and ordering the results in descending order. The answer for this question is the United States of America, with 7524410 reviews.


In [18]:
##QUESTION 3

In [19]:
result = secure_graph.run("""
    MATCH (uCTJ:Username{user_name:'CTJman'})-[mCTJ:MADE]->(reCTJ:Reviews)-[rCTJ:ABOUT]->(b:Beers)<-[r:ABOUT]-(re:Reviews)<-[m:MADE]-(u:Username)
    RETURN u.user_name,count(re) as cnt
    ORDER BY cnt DESC
    LIMIT 1
""").data()
pprint(result)

[{'cnt': 1428, 'u.user_name': 'acurtis'}]


In [20]:
R_edis.set("GROUP8:3", 'acurtis')

True

In this question the process made was connecting the Username to Reviews, and Review to Beers through the defined connection between files, in this first part of the **MATCH** was already defined that the Username wanted was "CTJman". Then we proceed to make the same connection in the second part of the MATCH but this time without any restriction. Having this MATCH build we move on for the **RETURN** part of the code and we specify that we wanted to return the Username(x) and the count of reviews connects to this Username(x). In this way we are counting the reviews that Username(x) has done in the same Beers as "CTJman" thus reaching our conclusion **Ordering by the count** and putting an **Limit of 1**.

In [21]:
##QUESTION 4

In [22]:
result = secure_graph.run("""
        MATCH (b:Beers)-[ba:BREWED_AT]->(br:Breweries)-[f:FROM]->(c:Country)
        WHERE c.country_digit='PT'
        RETURN br.name,count(b) as cnt
        ORDER BY cnt DESC
        LIMIT 1
""").data()
pprint(result)

[{'br.name': 'Dois Corvos Cervejeira', 'cnt': 40}]


In [23]:
R_edis.set("GROUP8:4", 'Dois Corvos Cervejeira', )

True

In this question the process made was connection in the **MATCH** the Beers to Brewery, and Brewery to Country. These connections were already defined through affiliations. The next step was adding a condition **WHERE** to specify that the Country from which we want information, in this case **'PT'**. So, in the next clause, the **RETURN** one, we defined that we wanted the Brewery name and the count of the Beers. In the last stage we just **Order by the count** and with this we can see the Portuguese Breweries that produce most beers.

In [24]:
##QUESTION 5

In [25]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)-[ba:BREWED_AT]->(br:Breweries)-[f:FROM]->(c:Country)
        WHERE c.country_digit='PT'
        RETURN b.name,count(re) as cnt
        ORDER BY cnt DESC
        LIMIT 1
""").data()
pprint(result)

[{'b.name': 'Super Bock', 'cnt': 391}]


In [26]:
R_edis.set("GROUP8:5", 'Super Bock')

True

The process in this question was very similar to one before we just added in the beginning on the first clause **MATCH** the Reviews connection to Beers using the ABOUT relationship that was pre-define. The difference between this question and the question before is that we have to **RETURN** the Beer name and the count of Reviews. Once we have this we can **Order by the count** and check the Portuguese Beers with the most reviews.

In [27]:
##QUESTION 6

In [28]:
result = secure_graph.run("""
        MATCH (br:Breweries)<-[ba:BREWED_AT]-(b:Beers)-[r:OF_TYPE]->(s:Style)
        WITH br, count(distinct s) as cnt
        RETURN avg(cnt)
""").data()
pprint(result)

[{'avg(cnt)': 10.595992516144603}]


In [29]:
round(result[0]['avg(cnt)'])

11

We understood that the output for this question should be the average of Styles given the number of all the breweries.
For this query we started by following the route that would connect Breweries to Style.
To do this we had to connect Beers to both, through BREWED_AT and OF_TYPE respectively, as this was their only common point.
Upon doing this we assigned a count of the distinct styles from the breweries to the variable count, returning their average. The value returned was 10.595 and since we can't have half a style we decided to round this output,leading to 11

In [30]:
R_edis.set("GROUP8:6", '11')

True

In [31]:
##QUESTION 7

In [32]:
result = secure_graph.run("""
        MATCH (b:Beers)-[ba:BREWED_AT]->(br:Breweries)
        WHERE toFloat(b.abv)>=20
        RETURN br.name,toFloat(b.abv),count(br.name)
        ORDER BY toFloat(b.abv) DESC, count(br.name) DESC
        LIMIT 1
""").data()
pprint(result)

[{'br.name': '1648 Brewing Company Ltd',
  'count(br.name)': 2,
  'toFloat(b.abv)': 100.0}]


In [33]:
R_edis.set("GROUP8:7", '1648 Brewing Company Ltd')

True

For the following query we started with a simple connection from Beers to Breweries, which was accomplished with BREWED_AT.
After this simple query we ordered our output by abv and after some research online found that the abv values were too high for beers. As such we defined 20 as an acceptable threshold for the strongest beers, as this only reflected a tiny amount of data.
After this decision we included a WHERE condition to filter out beers with an abv lower than 20, we had to include a toFloat() function as well, given the fact that the abv value was not a numeric value but rather a string.
With this condition in place, we returned the brewery's name alongside the abv of one of the latter beers and the number of times the brewery had a beer of the same abv. We implemented the latter as we found out that we had beers tied in 1st place, so decided on using this as a tiebreaker. 
With a simple ORDER BY the abv value, followed by a second ORDER BY the count of the number of times the brewery had a beer of the same abv and a LIMIT of 1 we found our strongest brewery, with it being 1648 Brewing Company LTD, which had 2 beers with an abv of 100.

In [34]:
##QUESTION 8

In [35]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)
        WHERE toFloat(re.smell) IS NOT NULL AND toFloat(re.look) IS NOT NULL
        WITH b,re,toFloat(re.smell) as Aroma, toFloat(re.look) as Appearance
        WITH b,re,(Aroma+Appearance)/2 as metric
        RETURN b.name, avg(metric) , count(re) 
        ORDER BY avg(metric) DESC, count(re) DESC
        LIMIT 1 
""").data()
pprint(result)

[{'avg(metric)': 5.0, 'b.name': 'Big Hefe', 'count(re)': 5}]


In [36]:
R_edis.set("GROUP8:8", 'Big Hefe')

True

For the following query we started with a simple connection from Reviews to Beers, which was accomplished through ABOUT.
Our first thought was to use the average of appearance value and the aroma value as the decider for the outcome. 
To do this we eliminated any possible null values in either variables, since there were reviews which only had the overall value but none of the variables that composed it.
Following this for an ease of usage we decided to appoint smell and look variables of the node to actual variables, with their conversion to floats done in the process.
For a prettier code, we used another WITH which made the calculation of the average mentioned beforehand.
We thus returned that output alongside the beer's name and the amount of reviews to serve as a tiebreaker, just like in the previous questions.
Following this we set two simple ORDER BY followed by a LIMIT 1 which resulted in the output Big Hefe, winning the tie-break with 5 reviews.


In [37]:
##QUESTION 9

In [38]:
result = secure_graph.run("""
        MATCH (re:Reviews)-[r:ABOUT]->(b:Beers)
        WHERE toFloat(re.overall) IS NOT NULL AND b.retired='f' AND (b.availability=' Spring' OR b.availability=' Year-round')
        WITH b,re,r,toFloat(re.overall) as Overall
        RETURN b.name,avg(Overall),count(r) as supercnt 
        ORDER BY avg(Overall) DESC,supercnt DESC
        LIMIT 3
""").data()
pprint(result)

[{'avg(Overall)': 5.0, 'b.name': 'Sunderland Mild', 'supercnt': 2},
 {'avg(Overall)': 5.0, 'b.name': 'Distorter Porter', 'supercnt': 1},
 {'avg(Overall)': 5.0, 'b.name': 'Scottish Perle', 'supercnt': 1}]


For question 9 we decided that we would use the overall value from the users recomendations to suggest the best beers.
The code structure is fairly similar to the one in the previous question. It does however include some extra clauses such as if the beverage is retired or not, since we don't want to suggest a beer hard to come by and by the same reason we chose the ones made in Spring, as it is the current season and as such we feel are the easier ones to get at the moment, in addition we also choose the ones year-round as these are always available

**Result** From the result of the query the beers recomended are Sunderland Mild,Distorter Porter and Scottish Perle